File: src/main/java/com/karma/hipgora/controller/chat/ChatController.java
package com.karma.hipgora.controller.chat;

import com.karma.hipgora.controller.requests.ReceivePublicChatRequest;
import com.karma.hipgora.controller.responses.ReceivePublicMessageResponse;
import com.karma.hipgora.model.chat.Chat;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.handler.annotation.SendTo;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@Slf4j
@RestController
@RequestMapping("/api/v1/chat")
public class ChatController {

    @Autowired
    private SimpMessagingTemplate simpMessagingTemplate;

    @MessageMapping("/message")
    @SendTo("/chatroom/public")
    public ReceivePublicMessageResponse receivePublicMessage(
            @Payload ReceivePublicChatRequest receivePublicChatRequest){
        return ReceivePublicMessageResponse.from(receivePublicChatRequest);
    }

    @MessageMapping("/private-message")
    public Chat receivePrivateMessage(@Payload Chat chat){
        simpMessagingTemplate.convertAndSendToUser(chat.getReceiver().toString(),"/private",chat);
        return chat;
    }
}

File: kotlin/backend/src/main/kotlin/com/karma/chat/controller/ChatController.kt
package com.karma.chat.controller


import com.karma.chat.controller.dto.BodyDto
import com.karma.chat.controller.dto.chat.SendPrivateChatMessageRequestDto
import com.karma.chat.controller.dto.chat.SendPublicChatMessageRequestDto
import com.karma.chat.domain.chat.OpenChatMessage
import com.karma.chat.domain.chat.PrivateChatMessage
import com.karma.chat.service.AccountService
import com.karma.chat.service.ChatService
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.messaging.handler.annotation.SendTo
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.web.bind.annotation.RestController

@RestController
class ChatController(
    private val accountService: AccountService,
    private val chatService: ChatService,
    private val messagingTemplate: SimpMessagingTemplate
) {
    @MessageMapping("/message")
    @SendTo("/chat/public")
    fun handlePublicMessage(@Payload req: SendPublicChatMessageRequestDto): BodyDto<OpenChatMessage> {
        val sender = accountService.extractUsernameFromJwt(req.jwt)
        val message = chatService.saveOpenChatMessage(content = req.content, sender = sender)
        return BodyDto(message = "success", data = message)
    }

    @MessageMapping
    @SendTo("/chat/private")
    fun handlePrivateMessage(@Payload req: SendPrivateChatMessageRequestDto): BodyDto<PrivateChatMessage> {
        val sender = accountService.extractUsernameFromJwt(req.jwt)
        val message =
            chatService.savePrivateChatMessage(content = req.content, sender = sender, receiver = req.receiver)
        return BodyDto(message = "success", data = message)
    }
}